### PR TITLE
Implement JAX dispatch for Split Op

### LINKED ISSUE #145: Missing JAX Implementation for `Split` `Op`
### Description

Hi everyone, since the update to Pymc 5.0 a code that used to run in pymc 4.0 started given me a lot of problems when using pm.sampling.jax.sample_numpyro_nuts. The model is as follows:

```python
def func1(x, alpha: float = 0.0, l_max: int = 12):
    cycles = [
        at.concatenate(
            [at.zeros(i), x[: x.shape[0] - i]]
        )
        for i in range(l_max)
    ]
    x_cycle = at.stack(cycles)
    w = at.as_tensor_variable([at.power(alpha, i) for i in range(l_max)])
    return at.dot(w, x_cycle)
```

```python
def func2(x, lam: float = 0.5):
    return (1 - at.exp(-lam * x)) / (1 + at.exp(-lam * x))

```

```python
with pm.Model() as asdr_model:
            asdr_model.add_coord('d', self.d, mutable = False)
            asdr_model.add_coord('ps', self.ps, mutable = False)
            asdr_model.add_coord('fourier_mode', np.arange(2*self.n_order), mutable = False)
            
            if add_trend:
                t_ = pm.Data(name=f"t", value= self.t, mutable = True)
            if add_seasonality:
                fourier_features_ = pm.Data(name=f"fourier_features", value= self.fourier_features, mutable = True)

            # --- priors ---

            ## intercept
            if add_trend:
                a = pm.Normal(name="a",
                              mu=priors_config['intercept']['mu'],
                              sigma=priors_config['intercept']['sigma'])
            else:
                a = pm.Normal(name="a",
                              mu=priors_config['intercept']['mu'],
                              sigma=priors_config['intercept']['sigma'])

            ## trend
            if add_trend:
                b_trend = pm.Normal(name="b_trend",
                                    mu=priors_config['b_trend']['mu'],
                                    sigma=priors_config['b_trend']['sigma'])
            ## seasonality
            if add_seasonality:
                b_fourier = pm.Laplace(name="b_fourier",
                                       mu=priors_config['b_seasonality']['mu'],
                                       b=priors_config['b_seasonality']['b'],
                                       dims="fourier_mode")

            ## standard deviation of the likelihood
            sigma = pm.HalfNormal(name='sigma',
                                  sigma=priors_config['sigma_likelihood']['sigma'])
            # degrees of freedom of the likelihood
            nu = pm.Gamma(name='nu',
                          alpha=priors_config['nu_likelihood']['alpha'],
                          beta=priors_config['nu_likelihood']['beta'])

            # Trend and seasonality determination
            if add_trend:
                trend = pm.Deterministic(name='trend', var = a + b_trend * t_, dims='date')
            if add_seasonality:
                seasonality = pm.Deterministic(
                    name='seasonality', var=pm.math.dot(fourier_features_, b_fourier), dims='date'
                )

            # --- data containers ---

            m_effects = []

            for ii, p in enumerate(self.ps):

                media_scaled_ = pm.Data(name=f"m_s_{p}", value=self.m_vs_s[:,ii], dims='date', mutable = True)

                ## func1 effect
                alpha = pm.Beta(name=f"alpha_{p}",
                                alpha=priors_config['alpha_adstock']['alpha'],
                                beta=priors_config['alpha_adstock']['beta'])
                ## func2 effect
                lam = pm.Gamma(name=f"lam_{p}",
                               alpha=priors_config['lambda_saturation']['alpha'],
                               beta=priors_config['lambda_saturation']['beta'])
                ## gaussian random walks standard deviation
                sigma_slope = pm.HalfNormal(name=f'sigma_slope_{p}',
                                            sigma=priors_config['sigma_slope']['sigma'])

                # --- model parametrization ---

                # Random walks for the betas of the media values
                slopes = pm.GaussianRandomWalk(
                    name=f'slopes_{p}',
                    sigma=sigma_slope,
                    init_dist=pymc.Normal.dist(
                        name=f'init_dist_{p}',
                        mu = priors_config['rw_initial_dist']['mu'],
                        sigma = priors_config['rw_initial_dist']['sigma']
                    ),
                    dims=['d'],
                )

                media_part1 = pm.Deterministic(
                    name=f"m_part1_{p}",
                    var=func1(x=m_s_, alpha=alpha, l_max=9),
                    dims=['d']
                )

                m_part1_part2 = pm.Deterministic(
                    name=f"m_part1_{p}",
                    var=func2(x=media_part1, lam=lam),
                    dims=['d'],
                )

                m_effect = pm.Deterministic(
                    name=f"m_part2_{p}",
                    var= pm.math.exp(slopes)*m_part1_part2,
                    dims = ['d']
                )

                m_effects.append(m_effect)


            # Final model
            obj_var = sum(m_effects)

            if add_trend:
                obj_var += trend
            else:
                obj_var += a
            if add_seasonality:
                obj_var += seasonality

            mu = pm.Deterministic(name="mu", var = obj_var, dims="d")


            # --- likelihood ---
            pm.StudentT(name="likelihood", nu=nu, mu=mu, sigma=sigma, observed=self.y_scaled, dims="d")

```

And the error is as follows:
First a lot of warnings:

```
/Users/leonardoalcivar/opt/anaconda3/envs/mmm_v3/lib/python3.11/site-packages/pymc/logprob/joint_logprob.py:167: UserWarning: Found a random variable that was neither among the observations nor the conditioned variables: [normal_rv{0, (0, 0), floatX, False}.0, init_dist_p]
```
And the error:
```
Cell In[30], line 280, in BayesianModel.train(self, model, iterations, chains)
    277 self.execution_time = now
    279 with model:
--> 280     self.model_trace = pm.sampling.jax.sample_numpyro_nuts(
    281         draws=iterations,
    282         chains=chains,
    283     )
    284     self.model_posterior_predictive = pm.sample_posterior_predictive(
    285         trace=self.model_trace
    286     )
    288 self.posterior_predictive_likelihood = (
    289     self.model_posterior_predictive.posterior_predictive[
    290         "likelihood"
    291     ].stack(sample=("chain", "draw"))
    292 )

File ~/opt/anaconda3/envs/mmm_v3/lib/python3.11/site-packages/pymc/sampling/jax.py:576, in sample_numpyro_nuts(draws, tune, chains, target_accept, random_seed, initvals, model, var_names, progress_bar, keep_untransformed, chain_method, postprocessing_backend, idata_kwargs, nuts_kwargs)
    567 print("Compiling...", file=sys.stdout)
    569 init_params = _get_batched_jittered_initial_points(
    570     model=model,
    571     chains=chains,
    572     initvals=initvals,
    573     random_seed=random_seed,
    574 )
--> 576 logp_fn = get_jaxified_logp(model, negative_logp=False)
    578 nuts_kwargs = _update_numpyro_nuts_kwargs(nuts_kwargs)
    579 nuts_kernel = NUTS(
    580     potential_fn=logp_fn,
    581     target_accept_prob=target_accept,
    582     **nuts_kwargs,
    583 )

File ~/opt/anaconda3/envs/mmm_v3/lib/python3.11/site-packages/pymc/sampling/jax.py:116, in get_jaxified_logp(model, negative_logp)
    114 if not negative_logp:
    115     model_logp = -model_logp
--> 116 logp_fn = get_jaxified_graph(inputs=model.value_vars, outputs=[model_logp])
    118 def logp_fn_wrap(x):
    119     return logp_fn(*x)[0]

File ~/opt/anaconda3/envs/mmm_v3/lib/python3.11/site-packages/pymc/sampling/jax.py:109, in get_jaxified_graph(inputs, outputs)
    106 mode.JAX.optimizer.rewrite(fgraph)
    108 # We now jaxify the optimized fgraph
--> 109 return jax_funcify(fgraph)

File ~/opt/anaconda3/envs/mmm_v3/lib/python3.11/functools.py:909, in singledispatch.<locals>.wrapper(*args, **kw)
    905 if not args:
    906     raise TypeError(f'{funcname} requires at least '
    907                     '1 positional argument')
--> 909 return dispatch(args[0].__class__)(*args, **kw)

File ~/opt/anaconda3/envs/mmm_v3/lib/python3.11/site-packages/pytensor/link/jax/dispatch/basic.py:49, in jax_funcify_FunctionGraph(fgraph, node, fgraph_name, **kwargs)
     42 @jax_funcify.register(FunctionGraph)
     43 def jax_funcify_FunctionGraph(
     44     fgraph,
   (...)
     47     **kwargs,
     48 ):
---> 49     return fgraph_to_python(
     50         fgraph,
     51         jax_funcify,
     52         type_conversion_fn=jax_typify,
     53         fgraph_name=fgraph_name,
     54         **kwargs,
     55     )

File ~/opt/anaconda3/envs/mmm_v3/lib/python3.11/site-packages/pytensor/link/utils.py:740, in fgraph_to_python(fgraph, op_conversion_fn, type_conversion_fn, order, storage_map, fgraph_name, global_env, local_env, get_name_for_object, squeeze_output, **kwargs)
    738 body_assigns = []
    739 for node in order:
--> 740     compiled_func = op_conversion_fn(
    741         node.op, node=node, storage_map=storage_map, **kwargs
    742     )
    744     # Create a local alias with a unique name
    745     local_compiled_func_name = unique_name(compiled_func)

File ~/opt/anaconda3/envs/mmm_v3/lib/python3.11/functools.py:909, in singledispatch.<locals>.wrapper(*args, **kw)
    905 if not args:
    906     raise TypeError(f'{funcname} requires at least '
    907                     '1 positional argument')
--> 909 return dispatch(args[0].__class__)(*args, **kw)

File ~/opt/anaconda3/envs/mmm_v3/lib/python3.11/site-packages/pytensor/link/jax/dispatch/basic.py:39, in jax_funcify(op, node, storage_map, **kwargs)
     36 @singledispatch
     37 def jax_funcify(op, node=None, storage_map=None, **kwargs):
     38     """Create a JAX compatible function from an PyTensor `Op`."""
---> 39     raise NotImplementedError(f"No JAX conversion for the given `Op`: {op}")

NotImplementedError: No JAX conversion for the given `Op`: Split{2}
```

### PR DESCRIPTION
**Pull request by ricardoV94**
Closes #145

### CODE REVIEW

File: pytensor/link/jax/dispatch/tensor_basic.py [hunk 1]
```diff
@@ -68,6 +75,57 @@ def join(axis, *tensors):
...
+        warnings.warn(
+            "Split node does not have constant split positions. Jax implementation will likely fail"
+        )
+
+    def split(x, axis, splits):
+        if constant_axis is not None:
+            axis = constant_axis
+        if constant_splits is not None:
+            splits = constant_splits
+            cumsum_splits = np.cumsum(splits[:-1])
```
Comment (twiecki) [inline pytensor/link/jax/dispatch/tensor_basic.py:107]: Why use the np version here?

Comment (ricardoV94) [reply pytensor/link/jax/dispatch/tensor_basic.py:107]: Because the jax one leads to a tracedarray that can't be used in `jnp.split` anymore.

Comment (twiecki) [reply pytensor/link/jax/dispatch/tensor_basic.py:107]: Maybe add a comment that this is a fall-back.

**Review by twiecki — APPROVED** [review]

File: pytensor/link/jax/dispatch/tensor_basic.py [hunk 2]
```diff
@@ -68,6 +75,57 @@ def join(axis, *tensors):
...
+        if np.sum(splits) != len_along_axis:
+            raise ValueError(
+                f"The splits do not sum up to array length along axis: {len_along_axis}"
+            )
+        if np.any(splits < 0):
+            raise ValueError(
+                "Attempted to make an array with a negative number of elements"
+            )
+
+        # tensors could also be tuples, and in this case they don't have a ndim
```
Comment (ricardoV94) [inline pytensor/link/jax/dispatch/tensor_basic.py:123]:
```suggestion
```

**Comment by michaelosthege** [comment]
I'm not familiar enough with this to make a difference here. Don't wait for my review

File: pytensor/tensor/basic.py [hunk 3]
```diff
@@ -1925,30 +1925,18 @@ def make_node(self, x, axis, splits):
     def perform(self, node, inputs, outputs):
         x, axis, splits = inputs
```
Comment (ferrine) [inline pytensor/tensor/basic.py:1926]: Is variable axis a frequent use case? If not why not setting it in the `Op.__props__`?

Comment (ricardoV94) [reply pytensor/tensor/basic.py:1926]: It's as supported as variable axis in `Join`, for tho grad of which `Split` is used.

Comment (ricardoV94) [reply pytensor/tensor/basic.py:1926]: This shouldn't pose any problem when users use constant axis

**Pull request merged**